### ISSUE TITLE
FairPlay media doesn't comply with HLS Authoring Specification for Apple devices

### ISSUE THREAD

**Issue by vladimir-kazakov** [shaka-project/shaka-packager#328]
The specification is available here: https://developer.apple.com/library/content/documentation/General/Reference/HLSAuthoringSpec/Requirements.html#//apple_ref/doc/uid/TP40016596-CH2-SW1

In the section 13 (content protection requirements) it says:
> 13.1. Content protection of video and audio SHOULD follow the FairPlay Streaming (FPS) specification.
> 13.2. If content is encrypted with FPS the method MUST be SAMPLE-AES.
> 13.3. If content is encrypted with FPS  the key format MUST be "com.apple.streamingkeydelivery".
> 13.4. The IV attribute SHOULD NOT be used with FPS unless necessary for interoperability.
> 13.5. Content with HD resolutions SHOULD use HDCP-LEVEL of TYPE-0.
> 13.6. Content with greater than HD resolutions SHOULD use HDCP-LEVEL of TYPE-1.

I'm trying to create FairPlay media (HLS, MP4, cbcs) using Shaka Packager `aa4bb49-release`, using the following command:

```
packager \
  'in=In/720p-2000k.mp4,stream=video,init_segment=Out/720p-2000k-init.mp4,segment_template=Out/720p-2000k-$Number$.mp4,playlist_name=720p-2000k.m3u8' \
  --enable_raw_key_encryption --protection_scheme cbcs --clear_lead 0 --mp4_include_pssh_in_stream=false \
  --keys 'label=:key_id=11111111111111111111111111111111:key=22222222222222222222222222222222' --iv 33333333333333333333333333333333 \
  --segment_duration 2 --fragment_duration 2 \
  --hls_key_uri 'skd://11111111-1111-1111-1111-111111111111' --hls_master_playlist_output Out/index.m3u8
```

But the output doesn't meet all requirements from the specification.

* 13.1. I assume that it's met, because after a few changes in the HLS media playlist file, the media successfully plays using FairPlay.
* 13.2. Is met.
* 13.3. Is not met. The value of the `KEYFORMAT` attribute of the `EXT-X-KEY` tag is always `identity`. I don't see any way to change it.
* 13.4. Is not met. The `IV` attribute in the `EXT-X-KEY` is always set. I don't see any way to disable it.
* 13.5. Is not met. The `EXT-X-STREAM-INF` tag, in the HLS master playlist file, doesn't have the `HDCP-LEVEL` attribute. There's no way to specify it.
* 13.6. Is not met. See 13.5.

The expected behavior would be meeting all requirements by default, but it also would be useful to control all SHOULDs manually. Perhaps Shaka Packager could have FairPlay flags, as it has for PlayReady and Widevine?

**Comment by kqyang** [comment]
For 13.3 and 13.4, you can enable Fairplay mode by providing a Fairplay PSSH. We use system id 29 70 1F E4 3C C7 4A 34 8C 5B AE 90 C7 43 9A 47 to signal fairplay key system (https://forums.developer.apple.com/thread/6185). You can generate a fairplay pssh using https://github.com/google/shaka-packager/blob/master/packager/tools/pssh/pssh-box.py tool with the fairplay system id. 

I am not familiar with HDCP in Apple systems. Shouldn't the HDCP constraint be associated with the licenses / keys instead of enforced in playlists, which can be easily modified?

**Comment by vladimir-kazakov** [comment]
> For 13.3 and 13.4, you can enable Fairplay mode by providing a Fairplay PSSH.

Nice, it works. Perhaps documentation could mention that.

> I am not familiar with HDCP in Apple systems. Shouldn't the HDCP constraint be associated with the licenses / keys instead of enforced in playlists, which can be easily modified?

As the HLS standard says in the [section 4.3.4.2](https://tools.ietf.org/html/rfc8216#section-4.3.4.2):
> This attribute is advisory;

> The HDCP-LEVEL attribute is OPTIONAL.  It SHOULD be present if any content in the Variant Stream will fail to play without HDCP. Clients without output copy protection SHOULD NOT load a Variant Stream with an HDCP-LEVEL attribute unless its value is NONE.

So, it's a way to keep clients from playing streams that they won't be able to play. This attribute is not related to FairPlay. But you're right, FairPlay licenses may have the HDCP policy, but this is already completely another story.

**Comment by kqyang** [comment]
> Nice, it works. Perhaps documentation could mention that.

It is tracked in https://github.com/google/shaka-packager/issues/306. We just do not have time to update it yet :(.

For HDCP-LEVEL, the problem is that we do not know what is the HDCP policy carried in the license. There is also a risk that the license is updated but the playlist is not, so the HDCP information in the playlist cannot be fully trusted.

We can potentially adding new options for HDCP-LEVEL and leave the decisions to the users. That is not the same as Apple recommendation though:

> 13.5. Content with HD resolutions SHOULD use HDCP-LEVEL of TYPE-0.
13.6. Content with greater than HD resolutions SHOULD use HDCP-LEVEL of TYPE-1.

Do you know if Apple has a definition on "HD resolutions"?

**Comment by vladimir-kazakov** [comment]
> For HDCP-LEVEL, the problem is that we do not know what is the HDCP policy carried in the license.

The packager doesn't need to know it. HDCP-LEVEL is additional info for clients, so that they can avoid streams that they may not be able to play. This attribute is no different from RESOLUTION, for example, which also may be used for the same purpose (e.g. the client doesn't support 4K+ resolutions).

> There is also a risk that the license is updated but the playlist is not, so the HDCP information in the playlist cannot be fully trusted.

The recommendation is to use HDCP for higher-quality content. If the license suddenly stops requiring HDCP for such content, this seems suspicious. Perhaps it's a mistake. In this case, HDCP-LEVEL may act as a safeguard. Yes, it's a weak safeguard, because according to the HLS standard, this attribute is advisory, so some clients may choose to ignore it. But it's still a protection from potential mistakes.

In case a content provider understands all risks of not using HDCP and decides to create content against recommendations, there will be an option to explicitly disable HDCP.

> We can potentially adding new options for HDCP-LEVEL and leave the decisions to the users. That is not the same as Apple recommendation though

Having such option will be very useful. Even if recommended defaults won't be appropriate for some users in some specific cases, this option will still allow them to precisely configure HDCP however they want.

> Do you know if Apple has a definition on "HD resolutions"?

I didn't find any explanation, but there is an example playlist in [appendix](https://developer.apple.com/library/content/documentation/General/Reference/HLSAuthoringSpec/appendix.html#//apple_ref/doc/uid/TP40016596-CH9-SW10) section of the specification, which shows that only 1080p is considered to be "HD", based on the usage of HDCP-LEVEL attribute.

**Comment by kqyang** [comment]
We are planning to add a new stream descriptor field for HDCP to allow users to specify HDCP levels if want to. Will work on it after milestone 2.1.0. Add the task to backlog for now.

**Comment by vladimir-kazakov** [comment]
> Do you know if Apple has a definition on "HD resolutions"?

I asked them, and they updated their [HLS Authoring Specification](https://developer.apple.com/documentation/http_live_streaming/hls_authoring_specification_for_apple_devices?language=objc) with the following explanation: "HD is approximately the range of 720p to 1080p."

So, according to this info, if content is protected using FairPlay, content from 720p to 1080p should use HDCP Type 0, and higher resolutions - HDCP Type 1.

**Comment by kqyang** [comment]
@vladimir-kazakov That is good to know. Thanks for the update!
